### DISCUSSION TITLE
Find the location of objects in a selection

### DISCUSSION THREAD

**Discussion by rockerBOO** [fabricjs/fabric.js#8477] [Q&A]
For my use case I am saving when objects are being moved while selected. I also have an undo/redo system that can change what is in the selection, where the objects are.

The issues this brings up mostly has to do with how to find the location of the object in the selection and how to compare locations of the objects that may be in a selection to where they are on the canvas.

- Need to reconcile the canvas based on the new state stored externally.
- Go through each object that is different from the canvas, 
 -  if it's in a selection we need to calculate the selection transforms and the object transforms. 
 -    If they don't match then we need to move the location the object inside the selection to the new location. 
 
But at this point, I can't figure out how to transform these values. 

I'm trying something like the following, to convert the object top,left over. This doesn't help change the scale, skew, angle coordinates directly.

```
  const mat = selection.calcTransformMatrix();

  // get position of the object inside the selection group.
  const coords = fabric.util.transformPoint(
    obj.getPointByOrigin(obj.originX, obj.originY),
    mat
  );
```

I tried doing various attempts with mutlipleTransformMatrices and invertTransform, but I am all off base at figuring it out for this case.  I looked in how selection was disposing of groups, but it was applying it to the object and not have a transform that was easy to compare.

I also noticed groups are getting reworked and also includes how selection works. On fabric.js 5.2.1, and not sure if there are improvements that could make this easier in newer releases. 

---

So simply, I want to find the transforms for objects in selection out from the object and in from the canvas, so that I can compare and/or apply modifications based on outside state through reconciliation.

---

I have a sample codebase which tries to show the location of objects on the canvas based on the object position and the "correct" transforms to place it in the same spot. 

https://codepen.io/rockerboo/pen/gOKjgvK

Showing selecting 2 items and the orange boxes are not lining up.

https://user-images.githubusercontent.com/15027/204378158-cd5c0941-ed6b-419a-b51c-6a70a8178ae8.mp4

Thank you!

**Comment by ShaMan123** [comment]
Let me start by pointing out that your explanation isn't very clear.
Do you need a local pointer? Would that solve your issue?
If I understand what you what to achieve the group rewrite will do that for you. Group methods `add` and `remove` do all the heavy lifting of the transformation changes now (v6).
In general be vigilant regarding relative coordinate planes. A lot of confusion around that. In the new website I will dedicate a guide for it. You can look at previous discussions with the group label I think.
I suggest you read #7670

**Reply by ShaMan123** [reply]
Otherwise if you must stick to v5 you will have to hack your way around this, probably by killing the active selection and recreating it.

**Reply by rockerBOO** [reply]
Thank you, @ShaMan123! I admit my explanation was a bit unclear, but mostly I was lost and confused. But you confirmed some things, and I appreciate all the work you're putting into making groups easier to use.

I wanted true location (on the canvas) and the selection relative location and how to calculate between those.

`.calcTransformMatrix()` and `.calcOwnMatrix()` were the 2 things, and `calcTransformMatrix` is always the true canvas relative transform and `calcOwnMatrix` is always relative to the nearest parent (canvas, selection group or group).

And then retrofitting or poly filling this [`applyTransformsToObject`](https://github.com/fabricjs/fabric.js/blob/fe6d9df08c5ff9b8b5e24e7b314e89071e60a362/src/util/misc/objectTransforms.ts#L58) we can understand how the transform matrix is applied to the object.

In my case, I want to compare transforms and positioning. So I can apply these methods to my own values for comparison. 

And lastly, to get the "selection" relative locations of objects inside, we invert the selection transform and multiply the true canvas relative transform matrix.

```
    fabric.util.multiplyTransformMatrices(
        fabric.util.invertTransform(selection.calcTransformMatrix()),
        obj.calcTransformMatrix()
      )
```

And this will get our selection relative positioning and transforms. Which then we can use for comparison or for applying to an object. I made a [blog post with more examples](https://davelage.com/posts/fabric-selection-group-positioning/).

**Reply by rockerBOO** [reply]
Also for killing the selection and re-adding, this causes cases where the selection is rotated to have a jarring snap so this would be a last resort, possibly when adding/removing objects but not as much for object position in the selection as these relative positions can be discovered and compared to true locations now with these calculations. 

I tried looking into re-applying the selection transforms and undoing them on the objects when they are added but it was a little too far for me  to go at this time. Would be interesting to be able to restore selection transforms and positions and have that invert the transforms and positions on each object or something.

**Reply by ShaMan123** [reply]
Look into the util sendObjectToPlane and sendPointToPlane
